Video: Rework buffer management

ngphibang opened the pull request:
This PR reworks the buffer management in the video subsystem step by step to have a common buffer management framework.

It fixes the current issues of video buffer management and may help to facilitate https://github.com/zephyrproject-rtos/zephyr/issues/89858.

Still a work-in-progress. I will try to describe the change in more details and complete it gradually.

TODO:

- Support user-provided buffer (static buffer, specific memory type etc.). Discussion: With this, we can move share_multiheap to this type (?) so that each time a specific type of memory is required, no need to modify the core framework => not a good idea, still keep SMH in the framework.
- Discussion : Support driver's memory or not ? => no usecase envisioned for now
- Support m2m devices
- Decouple CONFIG_VIDEO_BUFFER_POOL_NUM_MAX (max buffers for the whole system including all drivers + user-provided buffers) and driver's min_buffer requirement (e.g. CSI needs at least 2 buffers to work while DCMI, smartdma needs only 1 buffer) and user's desire.

ngphibang committed:
drivers: video: Request buffers instead of allocating

Replace video_buffer_alloc() and video_buffer_aligned_alloc() APIs with the
new video_request_buffers() API. With this,

- We can request multiple buffers at once.

- Application does not need to keep the empty allocated buffer pointer
  as it can enqueue the buffer via its index.

- Application may allocate its own buffers and provide them to the
  framework.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented:
Intteresting! Instead of having to carry pointers around, one only says "I need 10 buffers", then to enqueue them, "`dev0` uses buffers 0, 1, 2; `dev1` uses number 3, 4, 5, 6...". Keeping track of which buffer is used is just an integer `uint8_t used_buffer_num;` and bootstrapping buffer allocation is a lot easier.

**Q1:** Is it important that the application knows exactly which buffer index was given to a device? If not, then enqueue could be `video_buf_grant(dev, num_of_buffers);` and let `dev` do the allocation as needed.

**Q2:** It seems it is not possible anymore to manually craft a video buffer. For instance, if video data arrives from the network and then goes through PXP to adapt it to a display, how to submit the buffer coming directly from the network into a `struct video_buffer`?

**Q3:** Would it make sense to start making use of `video_buf_release(vbuf)` to automate the recycling of buffers? This can simplify the application that does not have to manually keep track of where to queue a buffer back.
-> `video_enqueue()` is only to push data to an m2m device or a video sink (the driver relesese the buffer),
-> `video_dequeue()` is only to acquire new data from an m2m device or video source (the application releases the buffer).

**Q4:** In my impression, an application developer does not want to know how many bytes are allocated to video in total, but instead decide on a resolution and let all allocation be automated. The current approach is to open `prj.conf` and a calculator. Should we try to address this now?

ngphibang committed:
drivers: video: Introduce video interface device

The main device that interface with application on memory.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: video_buffer: Switch from FIFO queues to RTIO

In the current implementation, drivers need to hold 2 FIFO queues (for
capture devices) or 4 FIFO queues (for m2m devices) leading to boilerplate
code to handle them.

Switch to use RTIO to benefit the following advantages:
- Better performances: mpsc is lockless so we can reduce context switch
- video_dequeue() is common and can be done by the framework
- video_enqueue() now becomes rtio_sqe_acquire() + rtio_sqe_prep_read()
  + rtio_submit() where the fist two parts are common and can be done by
  the framework. Drivers just need to implement iodev_submit() API
- Application is nearly unchanged but a call to video_rtio_cqe_release()
  after done with the full buffer to queue it back to sqe.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>
Co-developed-by: Josuah Demangeon <me@josuah.net>

ngphibang commented:
@josuah Hope you don't mind. I implemented the needed stuffs to generalize your RTIO driver demo and integtated it in this buffer management rework. Effectively, RTIO serves as the backend that replaces the FIFO queues (and some more) but it does not overlap or conflict with some changes in this PR which mainly reworks the buffer allocation and interfacing with application.

The main idea in the RTIO generalization is 
    - Application side nearly unchanged, still call enqueue / dequeue() except a call to video_release_buf() after finishing with the full buffer to queue it back to sqe. RTIO is opaque to application.
    - Driver just needs to hold an io_q and deal a bit with the buffer from sqe in isr. No need to implement rtio_iodev_api's submit()
    - The video subsystem does everything needed for buffer management.
    - We need the video_interface struct in https://github.com/zephyrproject-rtos/zephyr/pull/88182

It doesn't work yet :). I have error at stream_start(). I will try to debug later.

josuah reviewed (commented):
Thank you for picking it up so early!
I will try with an i.MXRT1160 I have around.

Essential to get it working is probably just to call `__frame_done_cb()` whenever a frame is submitted on `io_q()`, which might require `api->iodev_submit`...

ngphibang commented on drivers/video/video_common.c:
```diff
@@ -99,6 +102,101 @@ void video_buffer_release(struct video_buffer *vbuf)
+	struct rtio_iodev_sqe *iodev_sqe = CONTAINER_OF(node, struct rtio_iodev_sqe, q);
+	struct rtio_sqe *sqe = &iodev_sqe->sqe;
+
+	if (sqe->op != RTIO_OP_RX) {
+		LOG_ERR("Invalid operation %d of length %u for submission %p", sqe->op,
+			sqe->rx.buf_len, (void *)iodev_sqe);
+		rtio_iodev_sqe_err(iodev_sqe, -EINVAL);
+		return NULL;
+	}
```
My understand is this is to get the buffer that is filled from the sqe, to do some check (buf type check or buf_addr check as in csi driver). The code is common for all driver so it could be refactored as a function ?

josuah replied:
This makes sense as a helper agreed!

Maybe something like this?

```
struct video_buffer *video_get_io_q_vbuf(struct mpsc *io_q)
{
	struct video_buffer *vbuf;

	struct mpsc_node *node = mpsc_pop(io_q);
	if (node == NULL) {
		return NULL;
	}

	struct rtio_iodev_sqe *iodev_sqe = CONTAINER_OF(node, struct rtio_iodev_sqe, q);
	struct rtio_sqe *sqe = &iodev_sqe->sqe;

	vbuf = sqe->userdata;

	__ASSERT_NO_MSG(vbuf != NULL);

	if ((vbuf->type == VIDEO_BUF_TYPE_OUTPUT && (sqe->op != RTIO_OP_RX)) ||
	    (vbuf->type == VIDEO_BUF_TYPE_INPUT && (sqe->op != RTIO_OP_TX))) {
		return vbuf;
	} else {
		LOG_ERR("Unsupported RTIO operation (%d) or video buffer type (%d)",
			sqe->op, vbuf->type);
		return NULL;
	}
}
```

Proposal:

- Use `video_get_io_q_vbuf` to sentence it out as "get I/O queue's `vbuf`".

- Let the caller do `rtio_iodev_sqe_err(iodev_sqe, -EINVAL);` so that user call it regardless if it is from `video_get_io_q_vbuf() == NULL` or elsewhere.

ngphibang commented on drivers/video/video_device.c:
```diff
@@ -20,3 +22,45 @@ struct video_device *video_find_vdev(const struct device *dev)
+	const struct device *dev = iodev_sqe->sqe.iodev->data;
+	struct video_interface *vi = video_find_idev(dev);
+
+	mpsc_push(vi->io_q, &iodev_sqe->q);
```
In your original code, video_iodev_submit() call to driver's iodev_submit() then in the driver it does mpsc_push(vi->io_q, &iodev_sqe->q);

The code seems the same for all driver so that's why I let the driver keep a io_q and video_iodev_submit() does mpsc_push() directly, no need for driver implements iodev_submit() API

josuah replied:
Ah right! This is because `video_sw_generator.c` constantly try to enqueue frames based on a timer, so there is no need for it... If a driver needs to be triggered on the first frame submission, then `iodev_submit()` needs to be implemented. If not then there is no use-case.

For instance `video_stream_start()` triggers the first frame query, then there is no need for a distinct `iodev_submit()` also called `_start()` in some RTIO implementation.

You can mark this resolved I think this will not stand in the way.

ngphibang commented on drivers/video/video_device.h:
```diff
@@ -16,13 +17,31 @@ struct video_device {
 
+struct video_interface {
+	const struct device *dev;
+	struct mpsc *io_q;
```
It seems it's exactly what is done here. Driver jeep a struct mpsc io_q and initialize it in driver's init. RTIO's iodev does not hold iodev_sq anymore

ngphibang commented on drivers/video/video_common.c:
```diff
@@ -99,6 +102,101 @@ void video_buffer_release(struct video_buffer *vbuf)
+	}
+
+	return 0;
+}
```
Buffers are kept track of by index and type ?

josuah replied:
Right! `->index` type is fully user-defined, which allows user to organize a reference table in the application to store extra metadata about buffers.

I think this solves this point, feel free to fold this.

ngphibang commented on drivers/video/video_device.c:
```diff
@@ -20,3 +22,45 @@ struct video_device *video_find_vdev(const struct device *dev)
+static void video_iodev_submit(struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct device *dev = iodev_sqe->sqe.iodev->data;
+	struct video_interface *vi = video_find_idev(dev);
```
yes, that's right.

ngphibang commented:
I think I see the problem. Unlike the SW generator, the HW device like CSI needs to enqueue buffer to the HW, I blindly remove the enqueue() API.

ngphibang commented:
Just receive the 1st frame corrupted and then stopped.
![Uploading Corrupted.jpg…]()

ngphibang committed:
drivers: video: Move buffer management stuffs into a separate file

Move buffer management stuffs into video_buffer.c

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
> Q1: Is it important that the application knows exactly which buffer index was given to a device? If not, then enqueue could be video_buf_grant(dev, num_of_buffers); and let dev do the allocation as needed.

It already know and can track which buffer is enqueued.

> Q2: It seems it is not possible anymore to manually craft a video buffer. For instance, if video data arrives from the network and then goes through PXP to adapt it to a display, how to submit the buffer coming directly from the network into a struct video_buffer?

It was in TODO list and done now by commit `drivers: video: Add support for user-provided memory`

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -56,6 +56,29 @@ enum video_buf_type {
+};
+
+enum video_buf_memory {
+	/** buffer memory is on the video heap */
+	VIDEO_MEMORY_VIDEO = 1,
```
What about `VIDEO_MEMORY_POOL` to avoid a repetition?

josuah commented on drivers/video/video_buffer.c:
```diff
@@ -0,0 +1,229 @@
+	return 0;
+}
+
+void video_release_buf()
+{
+	/* Buffer will be re-queued thanks to RTIO_SQE_MULTISHOT */
+	rtio_cqe_release(&rtio, rtio_cqe_consume_block(&rtio));
+}
```
There is `video_buffer_release()` and `video_release_buf()`. To avoid confusion, how about `video_buffer_done()` or other name?

josuah commented on drivers/video/video_buffer.c:
```diff
@@ -0,0 +1,229 @@
+		return -EINVAL;
+	}
+
+	/* TODO: Buffer request is not cumulative. For each new request, free all the buffers
+	 * that are previously requested
+	 */
```
Data always comes out of something, be it application or video device, so at the time of allocation.

```
application  ---buffer--->   video_dev    /* input buffer, custom data from application */
video_dev    ---buffer--->   video_dev    /* output buffer, enqueue to video device first */
video_dev    ---buffer--->   application  /* output buffer, enqueue to video device first */
application  ---buffer--->   application  /* no need to use video APIs at all */
```

So there are really few different situations there, only 2:

- pass a custom pointer from the application and enqueue it to a video device for passing input data
- provide memory from the video memory pool and enqueue it to a video device for getting output data

Then, whether the buffer coming out of a video device is for the application or another video buffer is another concern (i.e. swap vbuf->type direction) that does not need to be solved during allocation.

If so, this eliminates a lot of user work:

```c
static unsigned int video_buf_allocated;

int video_alloc_output_buffer(uint8_t count, size_t size, const struct device *dev)
{
	if (video_buf_allocated + count > CONFIG_VIDEO_BUFFER_POOL_NUM_MAX) {
		return -EINVAL;
	}

	for (uint8_t i = 0; i < count; i++) {
		vbuf = video_buffer_aligned_alloc(size, CONFIG_VIDEO_BUFFER_POOL_ALIGN,
						  K_FOREVER);
		if (vbuf == NULL) {
			LOG_ERR("Failed to allocate video buffer %u/%u", i, count);
			return -ENOMEM;
		}

		vbuf->type = VIDEO_BUF_TYPE_OUTPUT;
		vbuf->index = video_buf_allocated;
		vbuf->size = size;
		vbuf->bytesused = 0;
		vbuf->state = VIDEO_BUF_STATE_DONE;

		ret = video_enqueue(dev, i);
		if (ret < 0) {
			LOG_ERR("Failed to queue buffer %u/%u to %s", i, count, dev->name);
			return ret;
		}

		video_buf_allocated++;
	}

	return video_buf_allocated;
}

int video_alloc_input_buffer(uint8_t *data, size_t size)
{
	if (video_buf_allocated + 1 > CONFIG_VIDEO_BUFFER_POOL_NUM_MAX) {
		return -EINVAL;
	}

	vbuf = &video_buf[video_buf_allocated];

	vbuf->type = VIDEO_BUF_TYPE_INPUT;
	vbuf->index = video_buf_allocated;
	vbuf->buffer = data;
	vbuf->size = size;
	vbuf->bytesused = size;
	vbuf->state = VIDEO_BUF_STATE_DONE;

	video_buf_allocated++;

	return video_buf_allocated;
}
```

Maybe you also planned to have a way to reset all allocations so that i.e. the pipeline can be reconfigured i.e. for a different format?

```c
int video_reset_buffers(void)
{
	/*
	 * Call RTIO API to cancel all operations, wait for completion of all jobs,
	 * then free all buffers
	 */
}
```

josuah replied:
This seems to allow making things automatic: for every video device, query capabilities, configure format, then for instance:

```c
int app_provision_video_dev(const struct device *dev)
{
	struct video_caps caps = {.type = VIDEO_BUF_TYPE_OUTPUT};
	struct video_format fmt = {.type = VIDEO_BUF_TYPE_OUTPUT};
	uint16_t line_count;

	ret = video_get_caps(dev, &caps);
	if (ret < 0) {
		return ret;
	}

	ret = video_get_caps(dev, &fmt);
	if (ret < 0) {
		return ret;
	}

	line_count = caps.min_line_count == LINE_COUNT_HEIGHT ? fmt.height : caps.min_line_count;

	ret = video_alloc_output_buffer(caps.min_vbuf_count, line_count * fmt.pitch, dev);
	if (ret < 0) {
		return ret;
	}

	return 0;
}
```

This allows to configure a video buffer then "allocate what this video device needs" and forget about it.

This works for `m2m` devices and source devices.

Destination-only devices do not need to worry about buffer allocation anyway, they receive already-allocated buffers.

How does this sound? Any catch I missed?

ngphibang replied:
> Is this really non-cumulative? How to support different input/output size? For instance compression or resizing? Then the input buffers and output buffers should have different size or memory is wasted.

We can always request buffers of different size each time we make a request (just cannot have different size in the same request as it does not make sense).

BTW, I thought to do as Linux, where it does not allow to cumulate requests, i.e. each request will free and invalidate buffers allocated by previous requests. But I realized now that Linux can do that because buffers is managed per device / driver so it make sense while in Zephyr buffers are managed centrally in the framework, so we need to allow to make multiple requests.

josuah commented on drivers/video/video_buffer.c:
```diff
@@ -0,0 +1,229 @@
+	return vbuf;
+}
+
+struct video_buffer *video_buffer_alloc(size_t size, k_timeout_t timeout)
+{
+	return video_buffer_aligned_alloc(size, sizeof(void *), timeout);
+}
```
Once the memory allocation is reworked, this function might be able to go away as never used?

josuah commented on drivers/video/video_buffer.c:
```diff
@@ -0,0 +1,229 @@
+	/* Allocate buffers */
+	for (uint8_t i = 0; i < count; i++) {
+		if (memory == VIDEO_MEMORY_VIDEO) {
+			buffer = video_buffer_aligned_alloc(size, CONFIG_VIDEO_BUFFER_POOL_ALIGN,
+							    K_FOREVER);
```
If SMH is used, `buffer` can be NULL as `timeout` is ignored.
```suggestion
			buffer = video_buffer_aligned_alloc(size, CONFIG_VIDEO_BUFFER_POOL_ALIGN, K_NO_WAIT);
			if (buffer == NULL) {
				return -ENOMEM;
			}
```

Modification applied above

ngphibang replied:
I think timeout should be a parameter in the function `video_request_buffers(uint8_t count, size_t size, k_timeout_t timeout)`.

PS: And type is not needed.

josuah commented on drivers/video/video_buffer.c:
```diff
@@ -0,0 +1,229 @@
+
+LOG_MODULE_REGISTER(video_buffer, CONFIG_VIDEO_LOG_LEVEL);
+
+#if defined(CONFIG_VIDEO_BUFFER_USE_SHARED_MULTI_HEAP)
+#include <zephyr/multi_heap/shared_multi_heap.h>
+
+#define VIDEO_COMMON_HEAP_ALLOC(align, size, timeout)                                              \
+	shared_multi_heap_aligned_alloc(CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE, align, size)
+#define VIDEO_COMMON_FREE(block) shared_multi_heap_free(block)
+#else
+K_HEAP_DEFINE(video_buffer_pool, CONFIG_VIDEO_BUFFER_POOL_SZ_MAX *CONFIG_VIDEO_BUFFER_POOL_NUM_MAX);
+#define VIDEO_COMMON_HEAP_ALLOC(align, size, timeout)                                              \
+	k_heap_aligned_alloc(&video_buffer_pool, align, size, timeout);
+#define VIDEO_COMMON_FREE(block) k_heap_free(&video_buffer_pool, block)
+#endif
```
Given the new allocation implementation does not use most of the parameters, this block could be simplified:

```suggestion
#if defined(CONFIG_VIDEO_BUFFER_USE_SHARED_MULTI_HEAP)

/* Wrappers around Shared MultiHeap (SMH) API */

#include <zephyr/multi_heap/shared_multi_heap.h>

#define VIDEO_COMMON_ALLOC(size)                                                                   \
	shared_multi_heap_aligned_alloc(CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE,                         \
					CONFIG_VIDEO_BUFFER_POOL_ALIGN, size)

#define VIDEO_COMMON_FREE(ptr) shared_multi_heap_free(ptr)

#else

/* Wrappers around k_heap() APIs */

K_HEAP_DEFINE(video_buffer_pool,                                                                   \
	      CONFIG_VIDEO_BUFFER_POOL_SZ_MAX * CONFIG_VIDEO_BUFFER_POOL_NUM_MAX);

#define VIDEO_COMMON_ALLOC(size)                                                                   \
	k_heap_aligned_alloc(&video_buffer_pool, CONFIG_VIDEO_BUFFER_POOL_ALIGN, size, K_NO_WAIT)

#define VIDEO_COMMON_FREE(block) k_heap_free(&video_buffer_pool, block)

#endif
```

josuah commented on drivers/video/video_buffer.c:
```diff
@@ -0,0 +1,229 @@
+void video_release_buf()
+{
+	/* Buffer will be re-queued thanks to RTIO_SQE_MULTISHOT */
+	rtio_cqe_release(&rtio, rtio_cqe_consume_block(&rtio));
```
`rtio_cqe_consume_block()` is already called in `video_dequeue()`, so it seems like we are popping the completion queue twice per buffer?

This mean we might need to expose the `struct rtio_cqe` so the user so the user can call `rtio_cqe_release(cqe)`?

josuah replied:
Resolved through recent commits. :+1:

josuah commented:
## `video_mcux_smartdma.c`

```
west build --board frdm_mcxn947/mcxn947/cpu0 --shield "dvp_20pin_ov7670 lcd_par_s035_8080"
samples/drivers/video/capture
```
![PXL_20250706_155456423](https://github.com/user-attachments/assets/849d7b57-50eb-48ba-868e-f69944cfa003)

josuah commented:
Removing auto-assigned reviewers for now to reduce notification spam, adding them back as soon as the work on video APIs is done.

josuah commented:
## `video_esp32_dvp.c`

```
west build --board esp32s3_eye/esp32s3/procpu samples/drivers/video/capture
```

![PXL_20250706_163611330](https://github.com/user-attachments/assets/068fb202-dbde-4737-97da-e981d01ac714)

ngphibang commented:
Nice ! Thank you for the follow-up ! I 've just picked up your commit. It works for the sw-generator ! For the CSI, I just naively did:

	`struct rtio_iodev_sqe *iodev_sqe = video_pop_io_q(&data->io_q);`

(as seen in the commit) and see that it is still not working yet. Sorry, I didn't take a look in RTIO yet, just did it naively.

josuah commented:
> About video_enqueue(), I have impression that we should do just [...] and driver will implement `.submit()` instead of `.enqueue()`.

Very good idea, I like it! This would look very similar to how RTIO works in other places of Zephyr, and help users understand it, and RTIO maintainers to understand how we use it.

ngphibang commented:
I removed driver's `.enqueue()` and replaced it by `.submit()` implementation. With this and the use of `video_pop_io_q()` in ISR, CSI and sw_generator is now working.

PS: Driver's implementation of `.submit()` is not mandatory, e.g. sw_generator does not need to implement it as common work is done in the framework.

ngphibang commented:
> Sorry I ran out of time while trying to load anything on my mimxrt1160_evk

Sorry for not inform you before working on it to avoid wasting your time, but I 've just work on this about 45 minutes ago ...

I will have to put this on hold for about the next 10 days due to some obligation. Thank you for the work !

josuah commented:
> Sorry for not inform you before working on it to avoid wasting your time, but I 've just work on this about 45 minutes ago ...

No time was lost as I spent it getting familiar with the i.MXRT ecosystem and tooling host-side mostly.

> I will have to put this on hold for about the next 10 days due to some obligation. Thank you for the work !

There is time before the end of feature freeze, and plenty of other drivers to convert.

What is left to do AFAIU:

- Decide the best way to handle memory allocation WRT these changes.
- Convert the drivers to the new API.

Thank you for the help!

josuah commented on drivers/video/video_common.c:
```diff
@@ -99,6 +102,101 @@ void video_buffer_release(struct video_buffer *vbuf)
+	video_buf[index].state = VIDEO_BUF_STATE_QUEUED;
+
+	/* RTIO submission */
+	struct rtio_iodev *ri = video_find_ridev(dev);
```
Resolving this thread as we can wait that we face the problem before planning a solution.

josuah commented:
## `video_stm32_dcmi.c`

```
# import config from capture_to_lvgl and then:
west build --board  mini_stm32h743 --shield weact_ov2640_cam_module samples/drivers/video/capture
```

![PXL_20250707_103334775](https://github.com/user-attachments/assets/2b1812ee-6717-4eee-9513-bb879676aa13)

available from this same branch here, which got rebased on top of the latest commits:
https://github.com/tinyvision-ai-inc/zephyr/pull/new/video_rework_buffer_management

Only DCMIPP and we should be done!
(and then, once it is merged, Renesas CEU from https://github.com/zephyrproject-rtos/zephyr/pull/92146)

ngphibang committed:
drivers: video: Enqueue and track internal buffers via index

Currently, the video_enqueue() API enqueues the whole external video
buffer container structure from the application and the driver stores
this container in its FIFO queue.

While it works in simple applications where the enqueued video_buffer
container persists for the whole program lifecycle, it does not work in
situations where we cannot keep this container, e.g. enqueuing a buffer
inside a function, the local variable will be destroyed when the function
returns and hence the buffer is no longer valid.

Video buffers can be tracked via their indices in the buffer pool.
Set the index field when buffers are allocated and enqueue the internal
buffer rather than the external one to fix the issue.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

avolmat-st commented on drivers/video/video_common.c:
```diff
@@ -70,6 +70,7 @@ struct video_buffer *video_buffer_aligned_alloc(size_t size, size_t align, k_tim
 	vbuf->buffer = block->data;
 	vbuf->size = size;
 	vbuf->bytesused = 0;
+	vbuf->state = VIDEO_BUF_STATE_DONE;
```
Just wondering, it depends on the application logic, but should we consider after allocation that the buffer state is DONE or should it be a state such as INIT for example ?
DONE would mean that it is filled with valid data.
Here, upon allocation, the buffer do not have valid data in it. bytesused is properly set to 0 so there is no chance the application would read bad value from it, so that's the reason why I do not really have strong arguments for a STATE_INIT.

josuah replied:
Would renaming `VIDEO_BUF_STATE_DONE` to  `VIDEO_BUF_STATE_OK` make sense then? Though it [differs a bit from Linux](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/buffer.html#buffer-flags).

avolmat-st commented on drivers/video/video_common.c:
```diff
@@ -101,26 +101,44 @@ void video_buffer_release(struct video_buffer *vbuf)
 	}
 
-	video_buf[buf->index].state = VIDEO_BUF_STATE_QUEUED;
+	video_buf[index].state = VIDEO_BUF_STATE_QUEUED;
```
state setting to be done after doing the enqueue since enqueue could fail

avolmat-st commented on drivers/video/video_common.c:
```diff
@@ -132,13 +147,28 @@ int video_enqueue(const struct device *dev, uint8_t index)
 	}
 
-	if (video_buf[index].state != VIDEO_BUF_STATE_DONE) {
+	if (video_buf[buf->index].type != buf->type ||
```
Maybe this is fixed in further commits but I am wondering if this check is ok when put in a chained devices pipeline. For example, in my current system I have two devices:
1. DCMIPP capture which output data, so I have output type which will get data from the camera pipeline.
2. JPEG Codec which is a M2M and has one input and one output. The M2M input is actually the same buffer as the DCMIPP output. So this unique buffer is sometimes queued / dequeue from the DMCIPP as output and something from the JPEG codec as input.
If the type is hardcoded in the video_buf table and the check is done when we enqueue, it seems it won't be possible to use a unique buffer between 2 devices. Did I understand it correctly ?

avolmat-st replied:
Oh actually seems I already asked this similar question before and @ngphibang replied me. Maybe I just haven't properly understood the code then.

ngphibang replied:
It's similar but not the same question :). The other one is the possibility of enqueuing the freshly dequeued buffer to another device (answer = yes). For this one, you are right, type shouldn't be checked here. Even more, type is not needed when requesting buffer. Initially, I did it by having the idea of supporting driver-provided buffer in mind, so each driver will have 1 or 2 buffer queues and the check is needed (buffer dequeued from one device will change its type when being queued into another device). But now, it seems user-provided and framework provided buffers are sufficient.

josuah commented on drivers/video/video_common.c:
```diff
@@ -113,6 +114,12 @@ int video_enqueue(const struct device *dev, struct video_buffer *buf)
+		return -EINVAL;
+	}
+
+	video_buf[buf->index].state = VIDEO_BUF_STATE_QUEUED;
```
Good point, maybe this needs to be set only if `ret = api->enqueue()` has `ret == 0`.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -600,7 +619,13 @@ static inline int video_dequeue(const struct device *dev, struct video_buffer **
+	ret = api->dequeue(dev, buf, timeout);
+	
+	if (ret == 0) {
+		(*buf)->state = VIDEO_BUF_STATE_DONE;
```
Here is the new `video_dequeue()` proposed:
https://github.com/nxp-upstream/zephyr/blob/bfc86f6ecb8f79238086b969abb9bbb35c149de3/drivers/video/video_buffer.c#L172-L175

Here is how it is used [for sensors](https://docs.zephyrproject.org/latest/hardware/peripherals/sensor/read_and_decode.html):

```c
			cqe = rtio_cqe_consume_block(&ctx);
			if (cqe->result != 0) {
				printk("I/O failed %d\n", cqe->result);
				return;
			}
```

So checking `cqe->result == 0` would give the result from the driver I/O operation, rather than just the "buffer pop" operation.

ngphibang committed:
drivers: video: video_common: Drop mem_block structure

The mem_block structure and the video_block array are not needed.
Drop them.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Support user-provided buffers

Add support for buffers allocated outside the video buffer heap.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
include: drivers: video: Some misc fixes

Some misc fixes:

- Use uint16_t for video_buffer's index
- Don't use enum in struct definition as it will be translated to int or
  other types depending on compiler and platform. Use uint8_t to save
  some bytes.
- Other minor fixes

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Kconfig: Fix VIDEO_BUFFER_POOL configs

The VIDEO_BUFFER_POOL_SZ_MAX and VIDEO_BUFFER_POOL_NUM_MAX configs are
not well defined. Rename and fix them.

Also, add range for VIDEO_BUFFER_POOL_NUM_BUFS which should correspond
to the range that video_buffer's index could cover, i.e. uint16_t range.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
Rebased on the latest main.

ngphibang commented:
Updates:

Added new commits:
- include: drivers: video: Some misc fixes
- drivers: video: video_common: Drop mem_block structure
- drivers: video: Kconfig: Fix VIDEO_BUFFER_POOL configs

Squashed RTIO related commits into one commit:
- drivers: video: video_buffer: Switch from FIFO queues to RTIO

Reworked all commits before RTIO (complete logics). TODO: need to complete these commits to pass CI

Tested user-provided static buffers:

```
	static uint8_t s_buffers[CONFIG_VIDEO_BUFFER_POOL_NUM_MAX][3686400] __nocache
		__aligned(CONFIG_VIDEO_BUFFER_POOL_ALIGN);

	struct video_buffer_request vbr = {
		.memory = VIDEO_MEMORY_EXTERNAL,
		.count = CONFIG_VIDEO_BUFFER_POOL_NUM_MAX,
		.size = 3686400,
	};

	vbuf->type = type;
	for (i = 0; i < vbr.count; i++) {
		vbuf->index = vbr.start_index++;
		vbuf->buffer = s_buffers[i];
		video_enqueue(video_dev, vbuf);
	}
```

RTIO part need to be optimized. I see **tearing effect** when tested on RT1170.

teburd reviewed (commented):
Some small comments based on our discussion in Discord

teburd commented:
I need to come back to this and plan to do so this year, along with looking at audio streaming and if we can also use RTIO for this (i2s/dmic).

This would at least potentially let a single thread do a full pipeline which is beneficial as I've many many times now that overuse of threads tends to be a net negative.

ngphibang commented:
> I need to come back to this and plan to do so this year, along with looking at audio streaming and if we can also use RTIO for this (i2s/dmic).
> 
> This would at least potentially let a single thread do a full pipeline which is beneficial as I've many many times now that overuse of threads tends to be a net negative.

We plan to start with net_buf for video buffer management before migrating to RTIO. Do you think the combination of net_buf and RTIO would be appropriate? Are there any potential conflicts or functional overlaps that you anticipate?

teburd commented:
> > I need to come back to this and plan to do so this year, along with looking at audio streaming and if we can also use RTIO for this (i2s/dmic).
> > This would at least potentially let a single thread do a full pipeline which is beneficial as I've many many times now that overuse of threads tends to be a net negative.
> 
> We plan to start with net_buf for video buffer management before migrating to RTIO. Do you think the combination of net_buf and RTIO would be appropriate? Are there any potential conflicts or functional overlaps that you anticipate?

RTIO really isn't about buffer management, its about having a queue of I/O like submission requests and completion events.

While we do have a memory block option involved with sensors, this is and should be entirely optional. That's there to at runtime allow for over burdening of some memory. The ideal is actually static allocation, but by statically allocating you committing more dedicated RAM. The benefit is you know there's always RAM there waiting to catch the samples you care about. With the block allocator you could I guess accept a worst case memory usage you want to allow for, and accept that if you run out of memory you will lose data.

Choosing net_buf as a structure you want to pass between an application and a video device might limit memory protected threads unless the buffer you are asking to read into or write from is very simple.

RTIO was written around this because you can pass these cheaply, without memcpy, between memory protected unprivileged threads to privileged code. As soon as there are branches on structures happening in privileged mode a memcpy will be required for those structures.

So if net_buf is used, this might be fine, but the linked list of buffers, and the structure around the data itself will need to be allocated and copied.

Does that make sense?

See https://docs.zephyrproject.org/latest/kernel/usermode/syscalls.html#verification-memory-access-policies for example which I've tried to adhere to when writing RTIO, which supports user mode.

I think this will only make more sense as time goes on and parts get more complex. We are already seeing many MMU capable devices in Zephyr, and video is likely something we would want to run in a memory protected context.

ngphibang commented:
> I need to come back to this and plan to do so this year, along with looking at audio streaming and if we can also use RTIO for this (i2s/dmic)

Thanks ! I think it will help as we experience some issues with audio buffer management when supporting dmic & i2s in libMP. I created this issue https://github.com/zephyrproject-rtos/zephyr/issues/107868 for us to discuss.
